testdrive: merge workflows into one ergonomic pile

benesch opened the pull request:
Merge all the testdrive workflows into a single `testdrive` workflow
that takes command-line flags. This makes it much more ergonomic to run
`testdrive` in various configurations from the command line.

I didn't see any downside to merging the Redpanda support back into the
main testdrive composition, because Redpanda is only launched when the
`--redpanda` option is supplied. This means that folks on M1 Macs can
use testdrive almost as easily as folks on Linux:

    ./mzcompose --dev run testdrive --redpanda foo.td

(@philip-stoev am I missing something here?)

I updated the documentation accordingly, and dropped in a few notes
about how unusably slow the Confluent Platform is on ARM machines.

Finally, I got rid of the `shell_eval` parameter to Testdrive, because
it was well beyond my capacity to understand. Instead, `testdrive` the
binary has learned to perform its own glob expansion, which covers 95%
of cases. The last 5% is the Redpanda nightly job, which needs a custom
workflow to handle filtering out the kafka-offset-times.td test file
which is not yet supported. This custom logic all lives in Python now,
though, so I think it's much easier to understand.

----

[Also a bunch of preceding cleanup commits; details in commit messages.]

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

Go commit by commit.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

philip-stoev commented on ci/nightly/pipeline.yml:
```diff
@@ -31,17 +31,17 @@
-        composition: redpanda
-        run: redpanda-testdrive
+        composition: testdrive
+        run: testdrive-redpanda-ci
```
I do not think there is a `testdrive-redpanda-ci` anymore, I guess you need `args` here with `--redpanda`

philip-stoev replied:
OH sorry my fault I spaced out.

philip-stoev commented on ci/nightly/pipeline.yml:
```diff
@@ -31,17 +31,17 @@
-        composition: redpanda
-        run: redpanda-testdrive
+        composition: testdrive
+        run: testdrive-redpanda-ci
```
Ditto here.

philip-stoev commented:
Item No 1. It seems that the new glob code does not handle relative paths. The `test/proxy` test attempts to run this:

```
$ docker-compose run testdrive-svc --aws-endpoint=http://localstack:4566 testdrive/avro-registry.td testdrive/esoteric/s3.td
```

but nothing gets run:
```
Creating proxy_testdrive-svc_run ... done
Configuration parameters:
    Kafka address: kafka:9092
    Schema registry URL: http://schema-registry:8081/
    materialized host: Tcp("materialized")
    error limit: 10
+++ testdrive completed successfully.
```

philip-stoev commented on src/testdrive/src/main.rs:
```diff
@@ -200,8 +200,27 @@ async fn main() {
+    if args.globs.is_empty() {
+        files.push(PathBuf::from("-"))
+    } else {
+        let walker = match GlobWalkerBuilder::from_patterns(".", &args.globs).build() {
```
Should this path only trigger if there are any wildcard characters being used? If no wildcard characters, just assume literal file names and use them verbatim, do not walk anything.

philip-stoev commented on src/testdrive/src/main.rs:
```diff
@@ -200,8 +200,27 @@ async fn main() {
+                    errors.push(e);
+                }
+            }
+        }
```
Now that we have control over globbing I think this should error out if the glob failed to match anything. Otherwise we risk test bugs where testdrive does not run anything without us noticing. Previously, failing to provide any .td files caused testdrive to hang waiting for input from STDIN, which was noticeable.

philip-stoev commented:
> (@philip-stoev am I missing something here?)

No, I moved the redpanda out at the time it was one gigantic yaml file, so it is OK to bring it back now that the thing is more manageable.

I can not see a downside -- the redpanda container is not even downloaded unless you actually attempt to start it up.

philip-stoev commented:
Item No 2. Specifying a non-existing test on the command line causes the entire thing to succeed silently.

```
./mzcompose run testdrive --redpanda no_such_test
...
$ docker-compose run testdrive-svc no_such_test
Creating testdrive_testdrive-svc_run ... done
Configuration parameters:
    Kafka address: kafka:9092
    Schema registry URL: http://schema-registry:8081/
    materialized host: Tcp("materialized")
    error limit: 10
+++ testdrive completed successfully.
```

philip-stoev commented:
Item No 3. I think the list of tests needs to be sorted prior to execution. This will match what the shell glob will do and is also important in certain tests where the order of .td files is significant and is set by prefixing each test with `10-`, `20-` etc.

philip-stoev commented:
Item No7. This does not run the tests in order:

```
./mzcompose run testdrive 'fetch*.td' 'array*.td'
```

array.td is being run somewhere in the middle of all the `fetch` tests. I think this is because `from_patterns` does not make an effort to keep the individual globs separate and ordered.

benesch commented:
Oof, thanks very much! I updated the glob handling to work as you suggest, and fixed at least one testdrive test that was attempting to pass multiple globs as a single argument. We'll see how this goes...

benesch commented:
All seems well, minus some pre-existing bugs in the Redpanda nightlies that I'll try to address tomorrow. Thanks for the reviews, everyone!
